File: python/run_modes.py
import os
from scheduler import EmployeeScheduler
from file_loader import load_from_file


def run_demo():
    print("\n" + "="*70)
    print("EMPLOYEE SCHEDULING SYSTEM - DEMO MODE")
    print("="*70 + "\n")
    
    scheduler = EmployeeScheduler()
    
    print("Adding employees...")
    employees = ["Alice", "Bob", "Charlie", "Diana", "Eve", "Frank", "Grace", "Henry"]
    for emp in employees:
        scheduler.add_employee(emp)
    
    print("\n" + "-"*70)
    print("Adding shift preferences...")
    print("-"*70 + "\n")
    
    scheduler.add_preference("Alice", "Monday", ["Morning", "Afternoon"])
    scheduler.add_preference("Alice", "Wednesday", ["Morning"])
    scheduler.add_preference("Alice", "Friday", ["Afternoon"])
    
    scheduler.add_preference("Bob", "Monday", ["Morning"])
    scheduler.add_preference("Bob", "Tuesday", ["Afternoon"])
    scheduler.add_preference("Bob", "Thursday", ["Evening"])
    scheduler.add_preference("Bob", "Saturday", ["Morning"])
    
    scheduler.add_preference("Charlie", "Tuesday", ["Morning", "Afternoon"])
    scheduler.add_preference("Charlie", "Wednesday", ["Afternoon"])
    scheduler.add_preference("Charlie", "Friday", ["Morning"])
    scheduler.add_preference("Charlie", "Sunday", ["Afternoon"])
    
    scheduler.add_preference("Diana", "Monday", ["Afternoon"])
    scheduler.add_preference("Diana", "Wednesday", ["Evening"])
    scheduler.add_preference("Diana", "Thursday", ["Morning"])
    scheduler.add_preference("Diana", "Saturday", ["Afternoon"])
    
    scheduler.add_preference("Eve", "Tuesday", ["Evening"])
    scheduler.add_preference("Eve", "Thursday", ["Afternoon"])
    scheduler.add_preference("Eve", "Friday", ["Evening"])
    scheduler.add_preference("Eve", "Sunday", ["Morning"])
    
    scheduler.add_preference("Frank", "Monday", ["Evening"])
    scheduler.add_preference("Frank", "Wednesday", ["Morning"])
    scheduler.add_preference("Frank", "Friday", ["Afternoon"])
    scheduler.add_preference("Frank", "Saturday", ["Evening"])
    
    scheduler.add_preference("Grace", "Tuesday", ["Morning"])
    scheduler.add_preference("Grace", "Thursday", ["Morning"])
    scheduler.add_preference("Grace", "Friday", ["Evening"])
    scheduler.add_preference("Grace", "Sunday", ["Afternoon"])
    
    scheduler.add_preference("Henry", "Monday", ["Morning"])
    scheduler.add_preference("Henry", "Wednesday", ["Afternoon"])
    scheduler.add_preference("Henry", "Thursday", ["Evening"])
    scheduler.add_preference("Henry", "Saturday", ["Morning"])
    
    scheduler.assign_shifts()
    scheduler.resolve_conflicts()
    scheduler.display_schedule()
    scheduler.get_statistics()


def run_interactive():
    print("\n" + "="*70)
    print("EMPLOYEE SCHEDULING SYSTEM - INTERACTIVE MODE")
    print("="*70 + "\n")
    
    scheduler = EmployeeScheduler()
    
    while True:
        try:
            num_employees = int(input("How many employees do you want to add? "))
            if num_employees > 0:
                break
            print("Please enter a positive number.")
        except ValueError:
            print("Please enter a valid number.")
    
    print("\nEnter employee names:")
    for i in range(num_employees):
        while True:
            name = input(f"  Employee {i+1}: ").strip()
            if name:
                scheduler.add_employee(name)
                break
            print("  Name cannot be empty.")
    
    # Add preferences
    print("\n" + "-"*70)
    print("Add shift preferences for each employee")
    print("Format: Enter shift numbers separated by spaces (1=Morning, 2=Afternoon, 3=Evening)")
    print("Or press Enter to skip a day")
    print("-"*70 + "\n")
    
    shift_map = {"1": "Morning", "2": "Afternoon", "3": "Evening"}
    
    for employee in scheduler.employees:
        print(f"\nPreferences for {employee}:")
        for day in scheduler.DAYS:
            pref_input = input(f"  {day} (1/2/3 or Enter to skip): ").strip()
            if pref_input:
                shift_nums = pref_input.split()
                shifts = [shift_map[num] for num in shift_nums if num in shift_map]
                if shifts:
                    scheduler.add_preference(employee, day, shifts)
    
    scheduler.assign_shifts()
    scheduler.resolve_conflicts()
    scheduler.display_schedule()
    scheduler.get_statistics()


def run_from_file(filename: str = None):
    print("\n" + "="*70)
    print("EMPLOYEE SCHEDULING SYSTEM - FILE IMPORT MODE")
    print("="*70 + "\n")
    
    scheduler = EmployeeScheduler()
    
    if not filename:
        script_dir = os.path.dirname(os.path.abspath(__file__))
        filename = os.path.join(os.path.dirname(script_dir), "sample_data.csv")
        print(f"Using default file: {filename}")
    
    if not load_from_file(scheduler, filename):
        print("\nFailed to load data from file. Exiting...")
        return
    
    if not scheduler.employees:
        print("\nNo employees loaded. Exiting...")
        return
    
    scheduler.assign_shifts()
    scheduler.resolve_conflicts()
    scheduler.display_schedule()
    scheduler.get_statistics()

File: java/src/runner/RunModes.java
package runner;

import scheduler.Scheduler;
import util.FileLoader;
import java.util.*;
import java.io.File;

public class RunModes {
    
    public static void runDemo() {
        System.out.println("\n" + "=".repeat(70));
        System.out.println("EMPLOYEE SCHEDULING SYSTEM - DEMO MODE");
        System.out.println("=".repeat(70) + "\n");
        
        Scheduler scheduler = new Scheduler();
        
        System.out.println("Adding employees...");
        String[] employeeNames = {"Alice", "Bob", "Charlie", "Diana", "Eve", "Frank", "Grace", "Henry"};
        for (String emp : employeeNames) {
            scheduler.addEmployee(emp);
        }
        
        System.out.println("\n" + "-".repeat(70));
        System.out.println("Adding shift preferences...");
        System.out.println("-".repeat(70) + "\n");
        
        scheduler.addPreference("Alice", "Monday", Arrays.asList("Morning", "Afternoon"));
        scheduler.addPreference("Alice", "Wednesday", Arrays.asList("Morning"));
        scheduler.addPreference("Alice", "Friday", Arrays.asList("Afternoon"));
        
        scheduler.addPreference("Bob", "Monday", Arrays.asList("Morning"));
        scheduler.addPreference("Bob", "Tuesday", Arrays.asList("Afternoon"));
        scheduler.addPreference("Bob", "Thursday", Arrays.asList("Evening"));
        scheduler.addPreference("Bob", "Saturday", Arrays.asList("Morning"));
        
        scheduler.addPreference("Charlie", "Tuesday", Arrays.asList("Morning", "Afternoon"));
        scheduler.addPreference("Charlie", "Wednesday", Arrays.asList("Afternoon"));
        scheduler.addPreference("Charlie", "Friday", Arrays.asList("Morning"));
        scheduler.addPreference("Charlie", "Sunday", Arrays.asList("Afternoon"));
        
        scheduler.addPreference("Diana", "Monday", Arrays.asList("Afternoon"));
        scheduler.addPreference("Diana", "Wednesday", Arrays.asList("Evening"));
        scheduler.addPreference("Diana", "Thursday", Arrays.asList("Morning"));
        scheduler.addPreference("Diana", "Saturday", Arrays.asList("Afternoon"));
        
        scheduler.addPreference("Eve", "Tuesday", Arrays.asList("Evening"));
        scheduler.addPreference("Eve", "Thursday", Arrays.asList("Afternoon"));
        scheduler.addPreference("Eve", "Friday", Arrays.asList("Evening"));
        scheduler.addPreference("Eve", "Sunday", Arrays.asList("Morning"));
        
        scheduler.addPreference("Frank", "Monday", Arrays.asList("Evening"));
        scheduler.addPreference("Frank", "Wednesday", Arrays.asList("Morning"));
        scheduler.addPreference("Frank", "Friday", Arrays.asList("Afternoon"));
        scheduler.addPreference("Frank", "Saturday", Arrays.asList("Evening"));
        
        scheduler.addPreference("Grace", "Tuesday", Arrays.asList("Morning"));
        scheduler.addPreference("Grace", "Thursday", Arrays.asList("Morning"));
        scheduler.addPreference("Grace", "Friday", Arrays.asList("Evening"));
        scheduler.addPreference("Grace", "Sunday", Arrays.asList("Afternoon"));
        
        scheduler.addPreference("Henry", "Monday", Arrays.asList("Morning"));
        scheduler.addPreference("Henry", "Wednesday", Arrays.asList("Afternoon"));
        scheduler.addPreference("Henry", "Thursday", Arrays.asList("Evening"));
        scheduler.addPreference("Henry", "Saturday", Arrays.asList("Morning"));
        
        scheduler.assignShifts();
        scheduler.resolveConflicts();
        scheduler.displaySchedule();
        scheduler.getStatistics();
    }
    
    public static void runInteractive() {
        System.out.println("\n" + "=".repeat(70));
        System.out.println("EMPLOYEE SCHEDULING SYSTEM - INTERACTIVE MODE");
        System.out.println("=".repeat(70) + "\n");
        
        Scanner scanner = new Scanner(System.in);
        Scheduler scheduler = new Scheduler();
        
        int numEmployees = 0;
        while (numEmployees <= 0) {
            System.out.print("How many employees do you want to add? ");
            try {
                numEmployees = Integer.parseInt(scanner.nextLine().trim());
                if (numEmployees <= 0) {
                    System.out.println("Please enter a positive number.");
                }
            } catch (NumberFormatException e) {
                System.out.println("Please enter a valid number.");
            }
        }
        
        System.out.println("\nEnter employee names:");
        for (int i = 0; i < numEmployees; i++) {
            String name = "";
            while (name.isEmpty()) {
                System.out.print("  Employee " + (i + 1) + ": ");
                name = scanner.nextLine().trim();
                if (name.isEmpty()) {
                    System.out.println("  Name cannot be empty.");
                }
            }
            scheduler.addEmployee(name);
        }
        
        System.out.println("\n" + "-".repeat(70));
        System.out.println("Add shift preferences for each employee");
        System.out.println("Format: Enter shift numbers separated by spaces (1=Morning, 2=Afternoon, 3=Evening)");
        System.out.println("Or press Enter to skip a day");
        System.out.println("-".repeat(70) + "\n");
        
        Map<String, String> shiftMap = new HashMap<>();
        shiftMap.put("1", "Morning");
        shiftMap.put("2", "Afternoon");
        shiftMap.put("3", "Evening");
        
        for (String employee : scheduler.employees) {
            System.out.println("\nPreferences for " + employee + ":");
            
            for (String day : Scheduler.DAYS) {
                System.out.print("  " + day + " (1/2/3 or Enter to skip): ");
                String prefInput = scanner.nextLine().trim();
                
                if (!prefInput.isEmpty()) {
                    String[] shiftNums = prefInput.split("\\s+");
                    List<String> shifts = new ArrayList<>();
                    
                    for (String num : shiftNums) {
                        if (shiftMap.containsKey(num)) {
                            shifts.add(shiftMap.get(num));
                        }
                    }
                    
                    if (!shifts.isEmpty()) {
                        scheduler.addPreference(employee, day, shifts);
                    }
                }
            }
        }
        
        scheduler.assignShifts();
        scheduler.resolveConflicts();
        scheduler.displaySchedule();
        scheduler.getStatistics();
        
        scanner.close();
    }
    
    public static void runFromFile(String filename) {
        System.out.println("\n" + "=".repeat(70));
        System.out.println("EMPLOYEE SCHEDULING SYSTEM - FILE IMPORT MODE");
        System.out.println("=".repeat(70) + "\n");
        
        Scheduler scheduler = new Scheduler();
        
        if (filename == null) {
            String currentDir = System.getProperty("user.dir");
            File parentDir = new File(currentDir).getParentFile();
            filename = new File(parentDir, "sample_data.csv").getAbsolutePath();
            System.out.println("No file specified. Using default: " + filename + "\n");
        }
        
        if (FileLoader.loadFromFile(scheduler, filename)) {
            if (!scheduler.employees.isEmpty()) {
                scheduler.assignShifts();
                scheduler.resolveConflicts();
                scheduler.displaySchedule();
                scheduler.getStatistics();
            } else {
                System.out.println("\nNo employees loaded.");
            }
        } else {
            System.out.println("\nFailed to load data from file.");
        }
    }
}
